### PR TITLE
feat: implement jsx-a11y addon

### PR DESCRIPTION
**Pull request by threequartersjohn**
This plugin implements an addon with our choice of rules from [`eslint-plugin-jsx-a11y`](https://github.com/evcohen/eslint-plugin-jsx-a11y#readme). 

So far, it explicitly excludes two rules:
- [`click-events-have-key-events`](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/click-events-have-key-events.md);
- [`media-has-caption`](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/media-has-caption.md).

Otherwise, all other rules are set as **errors**.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -43,6 +43,7 @@ Then enhance it with one or more **addons**:
 - `node` - If you are going to develop code for [NodeJS](nodejs.org)
 - `es-modules`: If you are going to use ES6 import & export instead of CommonJS or AMD (this rule skips root `[...].config.js` files to avoid ignoring this rule in common configuration files)
 - `babel-parser`: Use [babel-eslint](https://github.com/babel/babel-eslint) parser so that you may use language features that are not yet implemented in eslint (e.g.: dynamic imports)
+- `jsx-a11y`: If you're aiming for accessibility conformant development with JSX.
```
Comment (rafaelramalho19) [inline README.md]:
```suggestion
- `jsx-a11y`: If you are aiming for accessibility conformant development with JSX.
```

Just to be consistent :)

File: test/fixtures/rules/jsx-a11y/jsx-a11y.js [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    const bad = () => <html lang="PT" />;
+
+    // Succeed
+    const good = () => <html lang="pt" />;
+})();
+
+// '/media-has-caption': Enforces that <audio> and <video> elements must have a <track> for captions.
+(() => {
+    // Fail
+    const bad1 = () => <audio />;
```
Comment (rafaelramalho19) [inline test/fixtures/rules/jsx-a11y/jsx-a11y.js]: This is not failing because we disabled this rule.

File: test/fixtures/rules/jsx-a11y/jsx-a11y.js [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+    const bad = () => <input role="presentation" />;
+
+    // Succeed
+    const good = () => <input />;
+})();
+
+// '/no-noninteractive-element-interactions': Non-interactive elements should not be assigned mouse or keyboard event listeners.
+(() => {
+    // Fail
+    const bad = () => <ol onClick={ console.log() } onKeyDown={ console.log() } />;
```
Comment (rafaelramalho19) [inline test/fixtures/rules/jsx-a11y/jsx-a11y.js]: I personally don't agree with this. Let me explain why:

Given this use case:
```
<ol>
  <li /> * 10000 (10000 list items here)
</ol>
```

It's way faster and optimised to assign a onClick/onKeyDown event to the `ol` element and use the element present in the event (`event.target`).
It's not fast to assign 10000 event listeners to DOM elements.

Comment (rafaelramalho19) [reply test/fixtures/rules/jsx-a11y/jsx-a11y.js]: On the other hand (arguing with myself here 😄) this could be considered an edge-case and use a `eslint-ignore-no-noninteractive-element-interactions` inline. It looks uglier tho.

**Review by rafaelramalho19 — COMMENTED** [review]
Good job 👏

File: test/fixtures/rules/jsx-a11y/jsx-a11y.js (hunk 2 above)
Comment (threequartersjohn) [reply test/fixtures/rules/jsx-a11y/jsx-a11y.js]: Hmm, you're right, do you think i should document this in some other way? 

My reasoning for keeping it this way is, this file should reflect our options at any given point in time, and having the written example of a bad practice lacking the warning is, in my view, a proper reflection of our choices.

Also, though a bit more pedantically, i think we can say that we're testing for is the presence of the warnings, so, in effect, the fact that this is not failing means that the test is passing as it was expected to.

File: test/fixtures/rules/jsx-a11y/jsx-a11y.js (hunk 3 above)
Comment (threequartersjohn) [reply test/fixtures/rules/jsx-a11y/jsx-a11y.js]: I would also consider this an edge case, and would agree with using a ignore declaration. 

Or, at least, leave this for further consideration whenever we come across a situation like that.

File: test/fixtures/rules/jsx-a11y/jsx-a11y.js (hunk 2 above)
Comment (rafaelramalho19) [reply test/fixtures/rules/jsx-a11y/jsx-a11y.js]: If it's not enforced you should not document it :)

File: test/fixtures/rules/jsx-a11y/jsx-a11y.js [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+
+// '/no-distracting-elements': Enforce distracting elements are not used.
+(() => {
+    // Fail
+    const bad1 = () => <marquee />;
+    const bad2 = () => <blink />;
+
+    // Succeed
+    const good = () => <div />;
+})();
```
Comment (acostalima) [inline test/fixtures/rules/jsx-a11y/jsx-a11y.js]: Duplicated fixture?

File: rules/jsx-a11y.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        'jsx-a11y/aria-proptypes': 2,
+        // Enforce that elements with ARIA roles must use a valid, non-abstract ARIA role.
+        // https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/aria-role.md
+        'jsx-a11y/aria-role': 2,
+        // Enforce that elements that do not support ARIA roles, states, and properties do not have those attributes.
+        // https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/aria-unsupported-elements.md
+        'jsx-a11y/aria-unsupported-elements': 2,
+        // Enforce a clickable non-interactive element has at least one keyboard event listener.
+        // https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/click-events-have-key-events.md
+        'jsx-a11y/click-events-have-key-events': 0,
```
Comment (acostalima) [inline rules/jsx-a11y.js]: Does this rule cover scenarios such the one below?

```html
<ol>
   <li role="button" onClick={...} />
   ...
<ol>
```

If it does, then we should probably activate it, no?

**Review by acostalima — CHANGES_REQUESTED** [review]
Well done! 💪 ❤️

File: rules/jsx-a11y.js (hunk 5 above)
Comment (threequartersjohn) [reply rules/jsx-a11y.js]: This rule would cover that scenario, yes. However the current implementation of these rules would prevent even having a `button` role in an `<li>` element, because of the [`no-noninteractive-element-to-interactive-role`](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/no-noninteractive-element-to-interactive-role.md). 

Original conversation on this rule deemed that it should at most be a warning. 

Personally, i think the rule should be turned on only after also considering and making an effort at standardizing the expected key events.

File: test/fixtures/rules/jsx-a11y/jsx-a11y.js (hunk 4 above)
Comment (threequartersjohn) [reply test/fixtures/rules/jsx-a11y/jsx-a11y.js]: Yes :(

File: test/fixtures/rules/jsx-a11y/jsx-a11y.js (hunk 2 above)
Comment (threequartersjohn) [reply test/fixtures/rules/jsx-a11y/jsx-a11y.js]: @satazor @acostalima any opinions on this?

File: rules/jsx-a11y.js (hunk 5 above)
Comment (acostalima) [reply rules/jsx-a11y.js]: Could you please explain why are we enabling [`no-noninteractive-element-to-interactive-role`](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/no-noninteractive-element-to-interactive-role.md) as an error? There are valid use cases to use, e.g., interactive `li` tags. Following what you said above, I agree we should make this rule a warning at most.

Comment (acostalima) [reply rules/jsx-a11y.js]: In my opinion, [click-events-have-events](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/click-events-have-key-events.md) should be enabled as an error. Having a key event handler attached to an interactive tag is a basic accessibility requirement.

File: test/fixtures/rules/jsx-a11y/jsx-a11y.js (hunk 2 above)
Comment (acostalima) [reply test/fixtures/rules/jsx-a11y/jsx-a11y.js]: My question for the time being would be what's the reasoning to disable this rule? 🤔

**Review by acostalima — CHANGES_REQUESTED** [review]
See my latest comments.

File: packages/eslint-config-react-a11y/CHANGELOG.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
+# Change Log
```
Comment (satazor) [inline packages/eslint-config-react-a11y/CHANGELOG.md]: This should be deleted as it will be generated automatically

**Review by satazor — CHANGES_REQUESTED** [review]
It seems that treating all these new rules as errors seem pretty agressive.
I think that a good portion of them should be warnings. Thoughts?

**Comment by threequartersjohn** [comment]
> It seems that treating all these new rules as errors seem pretty aggressive.
> I think that a good portion of them should be warnings. Thoughts?

I'm ok with this. Though, we'll maybe need to revise this once a11y becomes a legal concern. At that point it would perhaps be better for most of them to be errors. But as it concerns this initial implementation, perhaps a "trial" period before a harsher implementation is a more effective way to go.

File: test/fixtures/rules/jsx-a11y/jsx-a11y.js (hunk 2 above)
Comment (acostalima) [reply test/fixtures/rules/jsx-a11y/jsx-a11y.js]: What are we going to do about this? 🤔

File: rules/jsx-a11y.js (hunk 5 above)
Comment (acostalima) [reply rules/jsx-a11y.js]: What are we going to do about this? 🤔

Comment (satazor) [reply rules/jsx-a11y.js]: @acostalima I might be wrong, but whenever you click enter or space in a clickable element, it will trigger a click event.

Comment (satazor) [reply rules/jsx-a11y.js]: If I stand correct, enabling this rule is too harsh for most situations. We just need to ensure that all elements with click listeners are focusable, so that users can use tab to select them.

**Comment by threequartersjohn** [comment]
Rules have now been reduced to warnings, with the exception of rules protecting against grammatical errors or otherwise any invalid `aria-` props (e.g. [`aria-props`](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/aria-props.md) or [`lang`](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/lang.md)).

**Review by nlfonseca — APPROVED** [review]
Looks Good 👍

File: rules/jsx-a11y.js (hunk 5 above)
Comment (acostalima) [reply rules/jsx-a11y.js]: > @acostalima I might be wrong, but whenever you click enter or space in a clickable element, it will trigger a click event.

@satazor Following the example in my comment above, I'm not so sure about elements where the `role="button"` is specified. Source: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#Required_event_handlers

Comment (satazor) [reply rules/jsx-a11y.js]: In those elements, the `tabIndex` should be specified, so that the elements are targetable via keyboard. @threequartersjohn do we enforce this via other rules?

File: test/fixtures/rules/jsx-a11y/jsx-a11y.js (hunk 2 above)
Comment (satazor) [reply test/fixtures/rules/jsx-a11y/jsx-a11y.js]: I'm fine enabling it but as a warning.

File: rules/jsx-a11y.js (hunk 5 above)
Comment (threequartersjohn) [reply rules/jsx-a11y.js]: No, actually we enforce that no `tabIndex` is specified in non interactive components ([`no-noninteractive-tabindex`](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/no-noninteractive-tabindex.md)), as we also enforce that no non-interactive component is given an interaction or interactive role ([`no-noninteractive-element-interactions`](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/no-noninteractive-element-interactions.md), [`no-noninteractive-element-to-interactive-role`](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/no-noninteractive-element-to-interactive-role.md)). 

In practice, there would be no situation where you'd need a `tabIndex` in an element like a `<div>` because you wouldn't make the `<div>` an interactive element. Please refer to the provided links for the reasoning behind each rule.

Comment (satazor) [reply rules/jsx-a11y.js]: @threequartersjohn I think you misunderstood my comment. I was referring to interactive elements and not non-interactive ones.

A div with a click event is an interactive element that should be focusable. That’s where tabindex come in: by defining a tabindex, it becomes focusable and reachable via keyboard (tab). Because a div with a tab index can be reached via keyboard, pressing enter on it will fire a click event (please see link mentioned by @acostalima).

Having that said, we should enforce that elements with onClick handlers have tabindex, except if they are buttons, anchors or other elements that are already focusable. So, if we use tabindex, having keyboard handlers is effectively unnecessary / redundant, hence making this rule unnecessary as well if we enforce tabindex.

Comment (threequartersjohn) [reply rules/jsx-a11y.js]: @acostalima after some tests, i'll turn on [click-events-have-events](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/click-events-have-key-events.md) as a warning.

This is because adding a `tabIndex`, a `button` role _and_ an `onclick` event, it still will not fire the `onclick` event through keyboard interaction as it does in buttons and similar elements. We should then guarantee that keyboard events are also present.

Comment (acostalima) [reply rules/jsx-a11y.js]: @threequartersjohn 👌

File: __snapshots__/index.spec.js.snap [hunk 7]
```diff
@@ -2043,6 +2043,17 @@ exports[`should pass on packages/eslint-config-react/test/fixtures/samples/react
 
 exports[`should pass on packages/eslint-config-react/test/fixtures/samples/react/contrived-old-es8.good.js 1`] = `Array []`;
 
+exports[`should pass on packages/eslint-config-react-a11y/test/fixtures/rules/jsx-a11y/jsx-a11y.js 1`] = `
+Array [
+  Object {
+    "column": 1,
+    "line": 1,
+    "rule": null,
```
Comment (satazor) [inline __snapshots__/index.spec.js.snap]: Hmmmm, this is strange. It should contain all the errors of the "Bad" sections of tests.

File: packages/eslint-config-react-a11y/package.json [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  "main": "index.js",
+  "files": [
+    "rules",
+    "lib"
+  ],
+  "peerDependencies": {
+    "eslint": "^6.0.0"
+  },
+  "dependencies": {
+    "@moxy/eslint-config-core": "^1.0.0",
```
Comment (satazor) [inline packages/eslint-config-react-a11y/package.json]: I think we may remove this dependency.

File: packages/eslint-config-react-a11y/rules/jsx-a11y.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
+'use strict';
+
+module.exports = {
+    'rules': {
+        // Enforce emojis are wrapped in and provide screen reader access.
+        // https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/accessible-emoji.md
```
Comment (satazor) [inline packages/eslint-config-react-a11y/rules/jsx-a11y.js]: Can we remove all these URLs in favor of just one at the top? See https://github.com/moxystudio/eslint-config/blob/master/packages/eslint-config-react/rules/react-hooks.js#L1 for an example how we are doing this in other packages.

File: packages/eslint-config-react-a11y/test/fixtures/rules/jsx-a11y/jsx-a11y.js [hunk 10]
```diff
@@ -0,0 +1,348 @@
+import React from 'react';
+
+// '/accessible-emoji': Enforce emojis are wrapped in and provide screen reader access.
```
Comment (satazor) [inline packages/eslint-config-react-a11y/test/fixtures/rules/jsx-a11y/jsx-a11y.js]: Can we keep these "separators" consistent with the other packages? see: https://github.com/moxystudio/eslint-config/blob/master/packages/eslint-config-react/test/fixtures/rules/react/react.js#L5

**Review by satazor — CHANGES_REQUESTED** [review]

File: rules/jsx-a11y.js (hunk 5 above)
Comment (satazor) [reply rules/jsx-a11y.js]: Resolving as this rule was activated as a warning.

File: test/fixtures/rules/jsx-a11y/jsx-a11y.js (hunk 2 above)
Comment (satazor) [reply test/fixtures/rules/jsx-a11y/jsx-a11y.js]: Resolving as this rule was activated as a warning.

File: __snapshots__/index.spec.js.snap (hunk 7 above)
Comment (threequartersjohn) [reply __snapshots__/index.spec.js.snap]: Yes, but as of the commit following this message i have tested this by manually running `eslint` on this file and it's reporting errors and warnings correctly, not sure if this is a problem with the IDE :/

Comment (satazor) [reply __snapshots__/index.spec.js.snap]: It seems they are good now.
